src/py/flwr_experimental/baseline/dataset/pytorch_cifar_partitioned.py

```diff
@@ -0,0 +1,218 @@
+import numpy as np
+import torch
+import torchvision
+import torchvision.transforms.functional as TF
+from torch import from_numpy
+from torchvision.utils import save_image
+from tqdm import tqdm
+
+from flwr_experimental.baseline.dataset.dataset import (
+    XY,
+    shuffle,
+    sort_by_label,
+    sort_by_label_repeating,
+)
+
+
+class Rot90Transform:
+    """Rotates image by a multiple of 90 degrees"""
+
+    def __init__(self, angles: List[int]):
+        self.angles = angles
+
+    def __call__(self, x: torch.Tensor) -> torch.Tensor:
+        angle = random.choice(self.angles)
+        return TF.rotate(x, angle)
+
+
+def augment_dataset(
+    original_dataset: torch.utils.data.Dataset,
+    augment_factor: int,
+    augment_transform: torchvision.transforms,
+    save_thumbnails_dir: str = "",
+) -> XY:
+    """Augments the dataset augment_transform
+
+    Dataset is augmented to len(original_dataset)*augment_factor
+    using augment_transform. The original dataset will be a subset
+    of the new one.
+
+    Parameters
+    ----------
+    original_dataset : torch.utils.data.Dataset
+        Original PyTorch Dataset to be augmented
+    augment_factor : int
+        Number of aumented copies generated from each original sample.
+        This will multiply the final length of the dataset by augment_factor.
+    augment_transform : torchvision.transforms
+        Composition of transforms used to augment the images.
+        Last transform in this composition does not need to be a ToTensor.
+    save_thumbnails_dir
+        Directory where to save thumbnails. One jpg image will be created for each original image
+        with a augment_factor thumbnails generated.
+
+    Returns
+    -------
+    XY
+        A tuple containing the new augmented dataset
+    """
+    if save_thumbnails_dir:
+        Path(save_thumbnails_dir).mkdir(parents=True, exist_ok=True)
+
+    combined_transform = torchvision.transforms.Compose(
+        [augment_transform, torchvision.transforms.ToTensor()]
+    )
+
+    augmented_imgs = np.empty(
+        (len(original_dataset), augment_factor, 3, 32, 32), dtype=np.float32
+    )
+    augmented_labels = np.empty((len(original_dataset), augment_factor), dtype=np.int32)
+
+    print("Generating augmented images...")
+    for idx, (img, label) in enumerate(tqdm(original_dataset)):
+        augmented_imgs[idx, 0, :, :, :] = torchvision.transforms.ToTensor()(img)
+        augmented_labels[idx, :] = label
+        for j in range(1, augment_factor):
+            augmented_imgs[idx, j, :, :, :] = combined_transform(img)
+        if isdir(save_thumbnails_dir):
+            tmp = [*augmented_imgs[idx, :, :, :, :]]
+            tmp2 = np.concatenate(tmp, axis=-1)
+            save_image(
+                from_numpy(tmp2), join(save_thumbnails_dir, f"cifar10_{idx}.jpg")
+            )
+
+    # Rearrange dataset. Every multiple of augment_factor is an original image.
+    X = augmented_imgs.reshape(-1, 3, 32, 32)
```

Comment:
Not sure but shouldn't this be: `X = augmented_imgs.reshape(-1, (3, 32, 32))` or am I getting something wrong?

Reply:
Tbh, I'm not sure if there is a practical difference. I expect it to generate a final tensor that is `(total number of images, 3, 32, 32)`.

Reply:
It might be backwards compatibility why its working although I might be wrong. Based on the docs

* https://pytorch.org/docs/stable/tensors.html#torch.Tensor.reshape
* https://numpy.org/doc/stable/reference/generated/numpy.reshape.html#numpy.reshape

I would suggest we follow the documented behavior and put the desired shape into a tuple.

Reply:
I see, the tuple, right? I'll change it to  `X = augmented_imgs.reshape((-1, 3, 32, 32))`

Reply:
solved in https://github.com/adap/flower/commit/5235c38e8259f44f47f609a2a1ee36bbda7c1864